lib/ex_unit/lib/ex_unit/runner.ex

```diff
@@ -332,6 +416,87 @@ defmodule ExUnit.Runner do
     :rand.seed(@rand_algorithm, {:erlang.phash2(module), :erlang.phash2(name), seed})
   end
 
+  defp register_test(pid, stats_pid) do
+    ExUnit.OnExitHandler.register(pid)
+    RunnerStats.register(stats_pid, pid)
+  end
+
+  defp register_test_module(pid) do
+    ExUnit.OnExitHandler.register(pid)
+  end
+
+  defp get_failure_counter(stats_pid) when is_pid(stats_pid),
+    do: RunnerStats.get_failure_counter(stats_pid)
+
+  defp increment_failure_counter(stats_pid, %struct{state: {tag, _}}, increment \\ 1)
+       when struct in [ExUnit.Test, ExUnit.TestModule] and tag in [:failed, :invalid],
+       do: RunnerStats.increment_failure_counter(stats_pid, increment)
+
+  # Takes care of the logic when the failure counter should be incremented,
+  # as well as stopping the suite if max_failures has been reached
+  defp process_failure(
+         process_failure_config,
+         %ExUnit.TestModule{state: {tag, _}, tests: tests} = test_module
+       )
+       when tag in [:failed, :invalid] do
+    %{config: config} = process_failure_config
+    failure_counter = increment_failure_counter(config.stats_pid, test_module, length(tests))
+
+    if max_failures_reached?(failure_counter, config.max_failures) do
+      max_failures_has_been_reached(config.manager, config.stats_pid, nil)
```

Comment:
Now that I see the code, I think killing all running tests will be harmful and potentially generate tons of logs. That's because if the test process has spawned other processes, killing the test process will be considered as a cascading failure. I think it is best just to not kill them and let them run their "course".

Also, don't we need to make a distinction between "max_failures" has just been reached and "max failures" has been surpassed? When it has just been reached, you still want to log that test but you want to stop the run. For subsequent cases, you don't want to log, nor continue the run.

Reply:
Just wondering if exiting like this would avoid the logging and also the cascading effect
```elixir
Process.exit(registered_pid, {:shutdown, :exit_max_failures_reached})
```

Reply:
So I have implemented your suggestion. Let me know if that works.
Regarding max failures reached and surpassed, we I don't think we need that, because I log the tests first and then I do `process_failures` which is based on the passed test, and just then I update the failure counter, if it's reached, I launch the `max_failures_have_been_reached` function, and any subsequent test will just be ignored and marked as :not_executed if they were still running